src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -318,6 +365,28 @@ void CodeGen::genAVX2Intrinsic(GenTreeHWIntrinsic* node)
             emit->emitIns_R_R_R(ins, emitTypeSize(TYP_SIMD32), targetReg, op1Reg, op2Reg);
             break;
         }
+
+        case NI_AVX2_Multiply:
+        {
+            op2Reg = op2->gtRegNum;
+            instruction ins;
+            switch (baseType)
+            {
+                case TYP_LONG:
+                    ins = INS_pmuldq;
```

Comment:
It looks to me that we should make all this table driven, otherwise we'll end up with a mountain of similar looking code.

Reply:
I just investigated class `Avx` that has about 75 method names (rather than overloads). Among these methods, about half (~35) are "simple" intrinsics (vector-only parameters) that can be easily table driven. `Avx2` is worse.
Actually, I do not oppose the table driven approach. I am still studying and exploring the implementation, so I am not sure what the code finally look like. For example, I am working on `Load`/`Store` recently. I found that CoreCLR/RyuJIT has had many code to deal with indirect memory access (e.g., `GT_IND`), reusing the code to matching SIMD HW intrinsics is non-trivial (Could you give some suggestions, by the way?).
Further more, high CQ of hardware intrinsic is not trivial neither. I am afraid that the table driven codgen may harm the subsequent optimization work.
In summary, I prefer to implement all (or most) intrinsics with the dirty and quick way at first. Then we can try to refactor it sometime.

Reply:
> I found that CoreCLR/RyuJIT has had many code to deal with indirect memory access (e.g., GT_IND), reusing the code to match SIMD HW intrinsics is non-trivial (Could you give some suggestions, by the way?).

Could you explain a bit more what the problem is? I know there are some complications in this area but I'm not sure what exactly are you referring to.

> Further more, high CQ of hardware intrinsic is not trivial neither.

Some examples would be useful. Right now the only obvious "problem" is how to map from `(intrinsicId, baseType)` to `instruction`.

To be clear - I do expect that there will be some special cases that might make the table driven approach problematic at times. But the fact that even something as basic as add/multiply require duplicated code is a bit concerning.

Reply:
> Could you explain a bit more what the problem is? I know there are some complications in this area but I'm not sure what exactly you are referring to.

First of all, what should the `Load`/`Store` IR look like? If we use `gtNewSimdHWIntrinsicNode(TYP_SIMD32, pointer, NI_AVX_Load, baseType, 32)`, is it possible to reuse the current indir code?

Reply:
I don't think it's feasible for load/stores to be intrinsics, they should be normal indirs having the appropriate type. Attempting to introduce a node that acts like and indir but it is not actually an indir sounds like a lot of trouble.

Also, it's not clear what purpose these load/store intrinsic serve. I expect code like `Vector128<int> v = Unsafe.Read<Vector128<int>>(p);` to work and generate an appropriate load instruction even if no intrinsic is used.

Only the aligned versions of the load/store intrinsics seems to provide value. And we need to figure out how to express those in IR. Probably we'll need to use a GTF flag to indicate that the address is aligned. Or reuse the existing `GTF_IND_UNALIGNED`.

cc @CarolEidt

Reply:
> Attempting to introduce a node that acts like and indir but it is not actually an indir sounds like a lot of trouble.
> Only the aligned versions of the load/store intrinsics seems to provide value. 

This concern is not only for `Load`/`Store`/`Aligned`/`Unaligned`, but we also have many other memory access intrinsics, e.g., 
```csharp
public static unsafe void ExtractVector128(byte* address, Vector256<byte> value, byte index);
public static unsafe void MaskStore(double* address, Vector256<double> mask, Vector256<ulong> source);
public static unsafe Vector128<int> GatherVector128(int* baseAddress, Vector128<int> index, byte scale);
public static unsafe Vector128<short> LoadAlignedNonTemporal(short* address);
```

Reply:
Hmm, let's see:
* `LoadAlignedNonTemporal` - this looks like an indir with a `GTF_NONTEMPORAL` flag. Hopefully there's room for a new flag.
* `ExtractVector128` - this seems to be `VEXTRACTI128` and this instruction can store to a register. This looks like it can be modeled as store indir + intrinsic. Though to be able to generate a single instruction the store indir operand (the intrinsic) would need to be contained and I'm not sure if that's possible right now. In the worst case we'll end up with two instructions, it's not that bad.

The other 2 are interesting, they always require memory operands. I suppose it's possible to use the same approach used for `ExtractVector128` with one condition - the JIT optimizer should never attempt to somehow eliminate the indir that the importer/morph creates. Most likely this means that these indir nodes needs to be excluded from CSE and if a local variable is behind the indir then the local variable must be made not enregistrable. And for MaskStore we need to deal with the containment issue.

There's sure going to be a bit of trouble to support these intrinsics by piggy backing on the existing indir nodes but it's likely to be simpler than attempting to hide indirs in intrinsic nodes.
Interlocked nodes already pull this stunt and that causes some problems, see #14557.

Reply:
> ExtractVector128 - this seems to be VEXTRACTI128 and this instruction can store to a register.

No, I intentionally exposed `Insert/ExtractVector128` as the memory-access intrinsic to provide the semantics of 16-byte load/store for AVX/AVX2 programming. 
That is a common pattern. Sometimes 32-byte alignment cannot guarantee (but 16-byte alignment can), so AVX 256-bit memory load/store would cause much more cache line split. Developers can use `Avx.Insert/ExtractVector128` to load/store data 128-bit by 128-bit to get better performance.